evap/grades/views.py

```diff
@@ -0,0 +1,174 @@
+    semester = get_object_or_404(Semester, id=semester_id)
+    course = get_object_or_404(Course, id=course_id)
+
+    grade_documents = GradeDocument.objects.filter(course=course)
+
+    template_data = dict(
+        semester=semester,
+        course=course,
+        grade_documents=grade_documents,
+        disable_if_archived="disabled=disabled" if semester.is_archived else "",
+        disable_breadcrumb_course=True,
+    )
+    return render(request, "grades_course_view.html", template_data)
+
+
+def helper_grade_upload(request, course, final_grades=False, instance=None):
+    created = True
+    if request.method == "POST":
+        if instance:
+            created = False
+        form = GradeDocumentForm(request.POST, request.FILES, instance=instance)
+        if form.is_valid():
+            instance = form.save(commit=False)
+            instance.last_modified_user = request.user
+            instance.course = course
+            if final_grades:
+                instance.type = GradeDocument.FINAL_GRADES
+            instance.save()
+            return True, created, form
+    else:
+        form = GradeDocumentForm(instance=instance, final_grades=final_grades)
+    return False, created, form
+
+
+@grade_publisher_required
+def upload_grades(request, semester_id, course_id):
+    semester = get_object_or_404(Semester, id=semester_id)
+    course = get_object_or_404(Course, id=course_id)
+
+    final_grades = request.GET.get('final', 'false') # default: preliminary grades
+    final_grades = {'true': True, 'false': False}.get(final_grades.lower()) # convert parameter to boolean
+
+    success, created, form = helper_grade_upload(request, course, final_grades=final_grades)
+
+    if success:
+        if created:
+            if not final_grades or course.state in ['evaluated', 'published']:
+                send_publish_notifications(grade_document_courses=[course])
+            elif course.state == 'reviewed':
+                course.publish()
+                course.save()
+                send_publish_notifications(grade_document_courses=[course], evaluation_results_courses=[course])
+
+        messages.success(request, _("Successfully uploaded grades."))
+        return redirect('grades:course_view', semester.id, course.id)
+    else:
+        template_data = dict(
+            semester=semester,
+            course=course,
+            form=form,
+            final_grades=final_grades,
+        )
+        return render(request, "grades_upload_form.html", template_data)
+
+
+@grade_downloader_required
+def download_grades(request, grade_document_id):
+    if not request.method == "GET":
+        return HttpResponseBadRequest()
+
+    grade_document = get_object_or_404(GradeDocument, id=grade_document_id)
+
+    # final grades can only be downloaded when the evaluation is finished
+    if grade_document.type == GradeDocument.FINAL_GRADES and grade_document.course.state not in ['evaluated', 'reviewed', 'published']:
```

Comment:
despite that, why would we hold back publishing the grades when evaluation is still running? surely that might influence the evaluation but whether grades are published or not shouldn't be our decision.